### PR TITLE
Add DCS & Runechat

### PR DESCRIPTION
**Pull request by ShadowLarkens**
DCS: https://github.com/ParadiseSS13/Paradise/pull/8901
DCS Improvements: https://github.com/ParadiseSS13/Paradise/pull/13594

Runechat PR: https://github.com/ParadiseSS13/Paradise/pull/14141

Added Runechat from /tg/, which shows speech messages from atoms and players alike over their head.
Additionally, added the whole datum component system from /tg/ as a side note to go along with it. It's just used for runechat to clean up messages when it's source gets deleted, for now, but you can [see all the uses that are possible here.](https://github.com/tgstation/tgstation/tree/master/code/datums/components)

There's a preference to turn it off here 

# Screenshots

Runechat messages have a 110 character limit. After that, it gets replaced with an ellipsis. That looks like this:

### CODE REVIEW

**Comment by ShadowLarkens** [comment]
DNM For now, I realize there's quite a bit missing from Paradise's PR

**Comment by ShadowLarkens** [comment]
So as it turns out, the missing stuff from Paradise's PR is intentional. At 150 players, /tg/'s full Runechat absolutely destroyed their server. This is an optimized version. I'll keep it that way, but I did go ahead and readd showing emotes and the preference to turn that on and off accordingly. Additionally, VV now allows you to arbitrarily add components and elements. Useful for debugging in the future.

No longer DNM, this is ready.

**Comment by AffectedArc07** [comment]
> So as it turns out, the missing stuff from Paradise's PR is intentional. At 150 players, /tg/'s full Runechat absolutely destroyed their server. This is an optimized version. I'll keep it that way, but I did go ahead and readd showing emotes and the preference to turn that on and off accordingly. Additionally, VV now allows you to arbitrarily add components and elements. Useful for debugging in the future.
> 
> No longer DNM, this is ready.

Can confirm. We run on a 9900k and even with that the GC was absolutely overwhelmed, so I cut features way back. Yes its low functionality compared to TGs implementation but we run this with zero issues

File: code/controllers/subsystems/runechat.dm [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  * after or adapted from the timer subsystem. Made by Bobbahbrown of /tg/station13
+  *
+  * Note that this has the same structure for storing and queueing messages as the timer subsystem does
+  * for handling timers: the bucket_list is a list of chatmessage datums, each of which are the head
+  * of a circularly linked list. Any given index in bucket_list could be null, representing an empty bucket.
+  *
+  * AA Note:
+  * One of the primary reasons for this is because each chatmessage has a timer attached to it, which is extra load on the GC
+  * At 150 population, the GC literally cannot keep up with processing 368,000 runechats and 368,000 extra timers in a 1 hour 30 minute round
+  * This also makes performance profiling a lot easier.
```
Comment (AffectedArc07) [inline code/controllers/subsystems/runechat.dm:21]: See this comment

File: code/modules/client/preferences.dm [hunk 2]
```diff
@@ -152,6 +152,11 @@ datum/preferences
 
 	var/list/volume_channels = list()
 
+	// Do we see runechat
+	var/runechat = TRUE
+	// Do we see runechat specifically for emotes?
+	var/runechat_emote = TRUE
```
Comment (AffectedArc07) [inline code/modules/client/preferences.dm:158]:
```suggestion
	/// Do we see runechat
	var/runechat = TRUE
	/// Do we see runechat specifically for emotes?
	var/runechat_emote = TRUE

```
Incase you ever do autodoc

File: code/modules/client/preferences_toggle_procs.dm [hunk 3]
```diff
@@ -460,3 +460,10 @@
 		SScharacter_setup.queue_preferences_save(prefs)
 
 	feedback_add_details("admin_verb","TBeSpecial") //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
+
+/client/verb/toggle_runechat()
+	set name = "Activate/Deactivate Runechat" // Called activate/deactivate to put it at the top of prefs for the sake of TMing
+	set category = "Preferences"
+	set desc = "Toggle runechat messages"
+	prefs.runechat = !prefs.runechat
+	to_chat(src, "You will [(prefs.runechat) ? "now see" : "no longer see"] floating chat messages.")
```
Comment (AffectedArc07) [inline code/modules/client/preferences_toggle_procs.dm:469]: May want to name this `Enable/Disable` for the sake of consistency, and remove the TM comment

**Comment by ShadowLarkens** [comment]
I'm too concerned about the performance issues runechat has to have this for now. I'll open a new PR with just DCS.

**Pull request closed without merging**